domain/application/state/containerimagemetadata.go

```diff
@@ -0,0 +1,163 @@
+`, storageKey, m)
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf(
+			"preparing select OCI image resource metadata statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, storageKey).Get(&m)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ResourceNotFound
+		} else if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf("selecting OCI image metadata: %w", err)
+	}
+
+	return application.ContainerImageMetadata{
+		UUID:         containermetadataresource.UUID(m.UUID),
+		StorageKey:   m.StorageKey,
+		RegistryPath: m.RegistryPath,
+		Username:     m.UserName,
+		Password:     m.Password,
+	}, nil
+}
+
+// Put stores container image metadata the database and returns its UUID.
+func (s *ContainerImageMetadataState) Put(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+	registryPath, userName, password string,
+) (containermetadataresource.UUID, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	uuid, err := containermetadataresource.NewUUID()
+	if err != nil {
+		return "", errors.Errorf("generating new UUID: %w", err)
+	}
+
+	m := containerImageMetadata{
+		UUID:         uuid.String(),
+		StorageKey:   resourceUUID.String(),
+		RegistryPath: registryPath,
+		UserName:     userName,
+		Password:     password,
+	}
+
+	stmt, err := s.Prepare(`
+INSERT INTO resource_oci_image_metadata_store (*)
+VALUES ($containerImageMetadata.*)
+`, m)
+	if err != nil {
+		return "", errors.Errorf("preparing insert OCI image resource metadata statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, m).Run()
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return "", errors.Errorf("inserting OCI image metadata: %w", err)
+	}
+	return uuid, nil
+}
+
+// Remove removes a container image metadata resource from storage.
+// applicationerrors.ResourceNotFound is returned if the resource does not
+// exist.
+func (s *ContainerImageMetadataState) Remove(ctx context.Context, uuid coreresource.ID) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	mUUID := containerImageMetadataStorageKey{StorageKey: uuid.String()}
+	stmt, err := s.Prepare(`
+DELETE FROM resource_oci_image_metadata_store 
+WHERE       storage_key = $containerImageMetadataStorageKey.storage_key
+`, mUUID)
+	if err != nil {
+		return errors.Errorf(
+			"preparing remove OCI image resource metadata statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var outcome sqlair.Outcome
+		err := tx.Query(ctx, stmt, mUUID).Get(&outcome)
+		if err != nil {
+			return err
+		}
+		if rows, err := outcome.Result().RowsAffected(); err != nil {
+			return err
+		} else if rows != 1 {
```

Comment:
**suggestion:** I think `rows == 0` is both more close to the intent and more reliable (if for some reason we can have several entry for a same storage key)